pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -82,135 +93,250 @@ func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *Co
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", job, err))
-		return
+func (tc *Controller) enqueueAfter(obj interface{}, after time.Duration) {
+	switch resource := obj.(type) {
+	case *batch.Job:
+		key, err := controller.KeyFunc(resource)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", resource, err))
+			return
+		}
+		tc.jobQueue.AddAfter(key, after)
+	case *v1.Pod:
+		key, err := controller.KeyFunc(resource)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", resource, err))
+			return
+		}
+		tc.podQueue.AddAfter(key, after)
 	}
+}
 
-	tc.queue.AddAfter(key, after)
+type queueType int
+
+const (
+	jobQueue queueType = iota
+	podQueue
+)
+
+func (tc *Controller) getQueue(queueType queueType) workqueue.RateLimitingInterface {
+	switch queueType {
+	case jobQueue:
+		return tc.jobQueue
+	case podQueue:
+		return tc.podQueue
+	default:
+		return nil
+	}
 }
 
-func (tc *Controller) worker() {
-	for tc.processNextWorkItem() {
+func (tc *Controller) jobWorker() {
+	for tc.processNextWorkItem(jobQueue) {
 	}
 }
 
-func (tc *Controller) processNextWorkItem() bool {
-	key, quit := tc.queue.Get()
+func (tc *Controller) podWorker() {
+	for tc.processNextWorkItem(podQueue) {
+	}
+}
+
+func (tc *Controller) processNextWorkItem(queueType queueType) bool {
+	var theQueue = tc.getQueue(queueType)
+	key, quit := theQueue.Get()
 	if quit {
 		return false
 	}
-	defer tc.queue.Done(key)
+	defer theQueue.Done(key)
 
-	err := tc.processJob(key.(string))
-	tc.handleErr(err, key)
+	err := tc.processResource(key.(string))
+	tc.handleErr(err, key, queueType)
 
 	return true
 }
 
-func (tc *Controller) handleErr(err error, key interface{}) {
+func (tc *Controller) handleErr(err error, key interface{}, queueType queueType) {
 	if err == nil {
-		tc.queue.Forget(key)
+		tc.getQueue(queueType).Forget(key)
 		return
 	}
 
-	utilruntime.HandleError(fmt.Errorf("error cleaning up Job %v, will retry: %v", key, err))
-	tc.queue.AddRateLimited(key)
+	utilruntime.HandleError(fmt.Errorf("error cleaning up Job/Pod %v, will retry: %v", key, err))
+	tc.getQueue(queueType).AddRateLimited(key)
 }
 
-// processJob will check the Job's state and TTL and delete the Job when it
-// finishes and its TTL after finished has expired. If the Job hasn't finished or
-// its TTL hasn't expired, it will be added to the queue after the TTL is expected
-// to expire.
+// processResource will check the Resource's [job / pod] state and TTL and
+// delete the resource when it finishes and its TTL after finished has expired.
+// If the resource hasn't finished or its TTL hasn't expired, it will be added
+// to the queue after the TTL is expected to expire.
 // This function is not meant to be invoked concurrently with the same key.
-func (tc *Controller) processJob(key string) error {
+func (tc *Controller) processResource(key string) error {
```

Comment:
We could process the wrong resource if a Job has the same key (namespace + name) as a Pod. How do we resolve that? The key needs to differentiate that.